File: Components/WinForms/VB/ChartsMainDemo/Modules/DataAnalysis/Trendlines.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Controls

Namespace DevExpress.XtraCharts.Demos.DataAnalysis
	Partial Public Class TrendlinesDemo
		Inherits ChartDemoBase2D
		Private trendlineBorder As ConstantLine = Nothing
		Private draggingTrendline As TrendLine = Nothing
		Private selectedTrendline As TrendLine = Nothing
		Private startValueLevelForNext As ValueLevel = ValueLevel.Close
		Private endValueLevelForNext As ValueLevel = ValueLevel.Close
		Private colorForNext As Color = Color.Empty
		Private extrapolateForNext As Boolean = True
		Private valueLevels() As ValueLevel = { ValueLevel.Low, ValueLevel.High, ValueLevel.Open, ValueLevel.Close }

		Private ReadOnly Property StartValueLevel() As ValueLevel
			Get
				Return valueLevels(cbeStarValueLevel.SelectedIndex)
			End Get
		End Property
		Private ReadOnly Property EndValueLevel() As ValueLevel
			Get
				Return valueLevels(cbeEndValueLevel.SelectedIndex)
			End Get
		End Property
		Private ReadOnly Property CurrentSeries() As Series
			Get
				Return If(chart.Series.Count > 0, chart.Series(0), Nothing)
			End Get
		End Property
		Private ReadOnly Property Diagram() As XYDiagram
			Get
				Return TryCast(chart.Diagram, XYDiagram)
			End Get
		End Property
		Private ReadOnly Property View() As FinancialSeriesViewBase
			Get
				Return If(CurrentSeries IsNot Nothing, TryCast(CurrentSeries.View, FinancialSeriesViewBase), Nothing)
			End Get
		End Property

		Public Overrides ReadOnly Property ChartControl() As ChartControl
			Get
				Return Me.chart
			End Get
		End Property

		Public Sub New()
			InitializeComponent()
		End Sub

		Private Function CalculateArgument(ByVal hitArgument As DateTime) As DateTime
			For i As Integer = 0 To CurrentSeries.Points.Count - 1
				If CurrentSeries.Points(i).DateTimeArgument < hitArgument Then
					Return If(i > 0, CurrentSeries.Points(i - 1).DateTimeArgument, CurrentSeries.Points(i).DateTimeArgument)
				End If
			Next i
			Return If(CurrentSeries.Points.Count > 0, CurrentSeries.Points(CurrentSeries.Points.Count - 1).DateTimeArgument, hitArgument)
		End Function
		Private Sub DeleteSelectedTrendline()
			If View IsNot Nothing AndAlso selectedTrendline IsNot Nothing Then
				View.Indicators.Remove(selectedTrendline)
				selectedTrendline = Nothing
			End If
			UpdateControls()
		End Sub
		Private Sub FillDefaultTrendlines()
			Dim trendline1 As New TrendLine("Trendline 2")
			trendline1.Point1.Argument = New DateTime(2016, 4, 15)
			trendline1.Point1.ValueLevel = ValueLevel.High
			trendline1.Point2.Argument = New DateTime(2016, 6, 12)
			trendline1.Point2.ValueLevel = ValueLevel.High
			trendline1.ExtrapolateToInfinity = False
			View.Indicators.Add(trendline1)

			Dim trendline2 As New TrendLine("Trendline 1")
			trendline2.Point1.Argument = New DateTime(2016, 6, 23)
			trendline2.Point1.ValueLevel = ValueLevel.High
			trendline2.Point2.Argument = New DateTime(2016, 10, 11)
			trendline2.Point2.ValueLevel = ValueLevel.High
			View.Indicators.Add(trendline2)

			Dim trendline3 As New TrendLine("Trendline 3")
			trendline3.Point1.Argument = New DateTime(2016, 3, 20)
			trendline3.Point1.ValueLevel = ValueLevel.Low
			trendline3.Point2.Argument = New DateTime(2016, 5, 13)
			trendline3.Point2.ValueLevel = ValueLevel.Low
			trendline3.ExtrapolateToInfinity = False
			View.Indicators.Add(trendline3)

			Dim trendline4 As New TrendLine("Trendline 4")
			trendline4.Point1.Argument = New DateTime(2016, 6, 27)
			trendline4.Point1.ValueLevel = ValueLevel.Low
			trendline4.Point2.Argument = New DateTime(2016, 10, 31)
			trendline4.Point2.ValueLevel = ValueLevel.Low
			View.Indicators.Add(trendline4)
		End Sub
		Private Sub FillComboBoxEditByValueLevels(ByVal comboboxEdit As ComboBoxEdit, ByVal selectedValueLevel As ValueLevel)
			comboboxEdit.Properties.Items.AddRange(valueLevels)
			comboboxEdit.SelectedItem = selectedValueLevel
		End Sub
		Private Sub chart_MouseDown(ByVal sender As Object, ByVal e As MouseEventArgs) Handles chart.MouseDown
			If View Is Nothing Then
				Return
			End If
			Dim hitInfo As ChartHitInfo = chart.CalcHitInfo(e.Location)
			If TypeOf hitInfo.Indicator Is TrendLine OrElse (Not hitInfo.InDiagram) OrElse e.Button <> MouseButtons.Left Then
				Return
			End If
			chart.SelectionMode = ElementSelectionMode.Single
			selectedTrendline = Nothing
			draggingTrendline = New TrendLine(View.Indicators.GenerateName("Trendline "))
			View.Indicators.Add(draggingTrendline)
			Dim coords As DiagramCoordinates = Diagram.PointToDiagram(e.Location)
			draggingTrendline.Point1.Argument = CalculateArgument(coords.DateTimeArgument)
			draggingTrendline.Point1.ValueLevel = startValueLevelForNext
			draggingTrendline.Point2.Argument = draggingTrendline.Point1.Argument
			draggingTrendline.Point2.ValueLevel = endValueLevelForNext
			draggingTrendline.ExtrapolateToInfinity = False
			draggingTrendline.Color = colorForNext
			trendlineBorder = New ConstantLine()
			trendlineBorder.AxisValue = draggingTrendline.Point1.Argument
			trendlineBorder.LineStyle.DashStyle = DashStyle.Dash
			trendlineBorder.LineStyle.Thickness = 1
			trendlineBorder.ShowInLegend = False
			View.AxisX.ConstantLines.Add(trendlineBorder)
			chart.Capture = True
			Cursor.Current = Cursors.SizeWE
			UpdateControls()
		End Sub
		Private Sub chart_MouseUp(ByVal sender As Object, ByVal e As MouseEventArgs) Handles chart.MouseUp
			If View Is Nothing Then
				Return
			End If
			If draggingTrendline IsNot Nothing Then
				draggingTrendline.ExtrapolateToInfinity = extrapolateForNext
				If draggingTrendline.Point1.Argument.Equals(draggingTrendline.Point2.Argument) Then
					View.Indicators.Remove(draggingTrendline)
				Else
					chart.SetObjectSelection(draggingTrendline)
				End If
				View.AxisX.ConstantLines.Remove(trendlineBorder)
				draggingTrendline = Nothing
			End If
			chart.Capture = False
			Cursor.Current = Cursors.Default
			UpdateControls()
		End Sub
		Private Sub chart_MouseMove(ByVal sender As Object, ByVal e As MouseEventArgs) Handles chart.MouseMove
			If View Is Nothing Then
				Return
			End If
			Dim hitInfo As ChartHitInfo = chart.CalcHitInfo(e.Location)
			If draggingTrendline IsNot Nothing AndAlso e.Button = MouseButtons.Left Then
				If hitInfo.InDiagram Then
					Dim coords As DiagramCoordinates = Diagram.PointToDiagram(e.Location)
					If (Not coords.IsEmpty) Then
						draggingTrendline.Point2.Argument = CalculateArgument(coords.DateTimeArgument)
						If CDate(draggingTrendline.Point1.Argument) > CDate(draggingTrendline.Point2.Argument) Then
							draggingTrendline.Point1.ValueLevel = EndValueLevel
							draggingTrendline.Point2.ValueLevel = StartValueLevel
						Else
							draggingTrendline.Point1.ValueLevel = StartValueLevel
							draggingTrendline.Point2.ValueLevel = EndValueLevel
						End If
						trendlineBorder.AxisValue = draggingTrendline.Point2.Argument
					End If
				End If
			ElseIf chart.SelectionMode = ElementSelectionMode.None Then
				chart.SelectionMode = ElementSelectionMode.Single
			End If
		End Sub
		Private Sub chart_KeyDown(ByVal sender As Object, ByVal e As KeyEventArgs) Handles chart.KeyDown
			If e.KeyCode = Keys.Delete Then
				DeleteSelectedTrendline()
			End If
		End Sub
		Private Sub btnDelete_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnDelete.Click
		   DeleteSelectedTrendline()
		End Sub
		Private Sub btnDeleteAll_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnDeleteAll.Click
			If View IsNot Nothing Then
				View.Indicators.Clear()
			End If
			selectedTrendline = Nothing
			UpdateControls()
		End Sub
		Private Sub cbeStarValueLevel_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbeStarValueLevel.SelectedIndexChanged
			If View IsNot Nothing Then
				If selectedTrendline Is Nothing Then
					startValueLevelForNext = StartValueLevel
				Else
					selectedTrendline.Point1.ValueLevel = StartValueLevel
				End If
			End If
		End Sub
		Private Sub cbeEndValueLevel_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbeEndValueLevel.SelectedIndexChanged
			If View IsNot Nothing Then
				If selectedTrendline Is Nothing Then
					endValueLevelForNext = EndValueLevel
				Else
					selectedTrendline.Point2.ValueLevel = EndValueLevel
				End If
			End If
		End Sub
		Private Sub cheExtrapolateToInfinity_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cheExtrapolateToInfinity.CheckedChanged
			If selectedTrendline Is Nothing Then
				extrapolateForNext = cheExtrapolateToInfinity.Checked
			Else
				selectedTrendline.ExtrapolateToInfinity = cheExtrapolateToInfinity.Checked
			End If
		End Sub
		Private Sub ceColor_CustomDisplayText(ByVal sender As Object, ByVal e As CustomDisplayTextEventArgs) Handles ceColor.CustomDisplayText
			If CType(e.Value, Color) = Color.Empty Then
				e.DisplayText = "Default"
			End If
		End Sub
		Private Sub ceColor_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs) Handles ceColor.EditValueChanged
			If selectedTrendline Is Nothing Then
				colorForNext = ceColor.Color
			Else
				selectedTrendline.Color = ceColor.Color
			End If
		End Sub
		Private Sub ceColor_Properties_ButtonClick(ByVal sender As Object, ByVal e As ButtonPressedEventArgs) Handles ceColor.Properties.ButtonClick
			If e.Button.Kind = ButtonPredefines.Delete Then
				ceColor.Color = Color.Empty
			End If
		End Sub
		Private Sub chart_BoundDataChanged(ByVal sender As Object, ByVal e As EventArgs) Handles chart.BoundDataChanged
			ChartControl.Animate()
		End Sub

		Protected Overrides Sub InitControls()
			MyBase.InitControls()
			FillDefaultTrendlines()
			FillComboBoxEditByValueLevels(cbeStarValueLevel, startValueLevelForNext)
			FillComboBoxEditByValueLevels(cbeEndValueLevel, endValueLevelForNext)
			ShowLabels = False
			CurrentSeries.DataSource = MarketPrices.GetAppleStockPrices()
		End Sub
		Protected Overrides Sub OnChartControlObjectHotTracked(ByVal sender As Object, ByVal e As HotTrackEventArgs)
			If Not(TypeOf e.Object Is TrendLine) Then
				If (Not chart.Capture) Then
					Cursor = Cursors.Default
				End If
				e.Cancel = True
			ElseIf draggingTrendline Is Nothing Then
				Cursor = Cursors.Hand
			Else
				e.Cancel = True
			End If
		End Sub
		Protected Overrides Sub OnChartControlObjectSelected(ByVal sender As Object, ByVal e As HotTrackEventArgs)
			If e.HitInfo.InIndicator Then
				selectedTrendline = TryCast(e.HitInfo.Indicator, TrendLine)
			Else
				selectedTrendline = Nothing
			End If
			If selectedTrendline Is Nothing Then
				e.Cancel = True
			End If
			UpdateControls()
		End Sub

		Public Overrides Sub UpdateControls()
			MyBase.UpdateControls()
			btnDelete.Enabled = selectedTrendline IsNot Nothing
			If View IsNot Nothing Then
				btnDeleteAll.Enabled = View.Indicators.Count <> 0
			End If
			If selectedTrendline Is Nothing Then
				cheExtrapolateToInfinity.Checked = extrapolateForNext
				cbeStarValueLevel.SelectedItem = startValueLevelForNext
				cbeEndValueLevel.SelectedItem = endValueLevelForNext
				ceColor.Color = colorForNext
			Else
				cheExtrapolateToInfinity.Checked = selectedTrendline.ExtrapolateToInfinity
				cbeStarValueLevel.SelectedItem = selectedTrendline.Point1.ValueLevel
				cbeEndValueLevel.SelectedItem = selectedTrendline.Point2.ValueLevel
				ceColor.Color = selectedTrendline.Color
			End If
		End Sub
	End Class
End Namespace

File: Components/WinForms/CS/ChartsMainDemo/Modules/DataAnalysis/Trendlines.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Controls;

namespace DevExpress.XtraCharts.Demos.DataAnalysis {
    public partial class TrendlinesDemo : ChartDemoBase2D {        
        ConstantLine trendlineBorder = null;
        TrendLine draggingTrendline = null;
        TrendLine selectedTrendline = null;
        ValueLevel startValueLevelForNext = ValueLevel.Close;
        ValueLevel endValueLevelForNext = ValueLevel.Close;
        Color colorForNext = Color.Empty;
        bool extrapolateForNext = true;
        ValueLevel[] valueLevels = new ValueLevel[] { ValueLevel.Low, ValueLevel.High, ValueLevel.Open, ValueLevel.Close };

        ValueLevel StartValueLevel { get { return valueLevels[cbeStarValueLevel.SelectedIndex]; } }
        ValueLevel EndValueLevel { get { return valueLevels[cbeEndValueLevel.SelectedIndex]; } }
        Series CurrentSeries { get { return chart.Series.Count > 0 ? chart.Series[0] : null; } }
        XYDiagram Diagram { get { return chart.Diagram as XYDiagram; } }
        FinancialSeriesViewBase View { get { return CurrentSeries != null ? CurrentSeries.View as FinancialSeriesViewBase : null; } }
        
        public override ChartControl ChartControl { get { return this.chart; } }

        public TrendlinesDemo() {
            InitializeComponent();
        }

        DateTime CalculateArgument(DateTime hitArgument) {            
            for (int i = 0; i < CurrentSeries.Points.Count; i++) {
                if (CurrentSeries.Points[i].DateTimeArgument < hitArgument)
                    return i > 0 ? CurrentSeries.Points[i - 1].DateTimeArgument : CurrentSeries.Points[i].DateTimeArgument;
            }
            return CurrentSeries.Points.Count > 0 ? CurrentSeries.Points[CurrentSeries.Points.Count - 1].DateTimeArgument : hitArgument;
        }
        void DeleteSelectedTrendline() {
            if (View != null && selectedTrendline != null) {
                View.Indicators.Remove(selectedTrendline);
                selectedTrendline = null;
            }
            UpdateControls();
        }
        void FillDefaultTrendlines() {
            TrendLine trendline1 = new TrendLine("Trendline 2");
            trendline1.Point1.Argument = new DateTime(2016, 4, 15);
            trendline1.Point1.ValueLevel = ValueLevel.High;
            trendline1.Point2.Argument = new DateTime(2016, 6, 12);
            trendline1.Point2.ValueLevel = ValueLevel.High;
            trendline1.ExtrapolateToInfinity = false;
            View.Indicators.Add(trendline1);

            TrendLine trendline2 = new TrendLine("Trendline 1");
            trendline2.Point1.Argument = new DateTime(2016, 6, 23);
            trendline2.Point1.ValueLevel = ValueLevel.High;
            trendline2.Point2.Argument = new DateTime(2016, 10, 11);
            trendline2.Point2.ValueLevel = ValueLevel.High;
            View.Indicators.Add(trendline2);

            TrendLine trendline3 = new TrendLine("Trendline 3");
            trendline3.Point1.Argument = new DateTime(2016, 3, 20);
            trendline3.Point1.ValueLevel = ValueLevel.Low;
            trendline3.Point2.Argument = new DateTime(2016, 5, 13);
            trendline3.Point2.ValueLevel = ValueLevel.Low;
            trendline3.ExtrapolateToInfinity = false;
            View.Indicators.Add(trendline3);

            TrendLine trendline4 = new TrendLine("Trendline 4");
            trendline4.Point1.Argument = new DateTime(2016, 6, 27);
            trendline4.Point1.ValueLevel = ValueLevel.Low;
            trendline4.Point2.Argument = new DateTime(2016, 10, 31);
            trendline4.Point2.ValueLevel = ValueLevel.Low;
            View.Indicators.Add(trendline4);
        }
        void FillComboBoxEditByValueLevels(ComboBoxEdit comboboxEdit, ValueLevel selectedValueLevel) {
            comboboxEdit.Properties.Items.AddRange(valueLevels);
            comboboxEdit.SelectedItem = selectedValueLevel;
        }
        void chart_MouseDown(object sender, MouseEventArgs e) {
            if (View == null)
                return;
            ChartHitInfo hitInfo = chart.CalcHitInfo(e.Location);
            if (hitInfo.Indicator is TrendLine || !hitInfo.InDiagram || e.Button != MouseButtons.Left)
                return;
            chart.SelectionMode = ElementSelectionMode.Single;
            selectedTrendline = null;
            draggingTrendline = new TrendLine(View.Indicators.GenerateName("Trendline "));            
            View.Indicators.Add(draggingTrendline);
            DiagramCoordinates coords = Diagram.PointToDiagram(e.Location);
            draggingTrendline.Point1.Argument = CalculateArgument(coords.DateTimeArgument);
            draggingTrendline.Point1.ValueLevel = startValueLevelForNext;
            draggingTrendline.Point2.Argument = draggingTrendline.Point1.Argument;
            draggingTrendline.Point2.ValueLevel = endValueLevelForNext;
            draggingTrendline.ExtrapolateToInfinity = false;
            draggingTrendline.Color = colorForNext;
            trendlineBorder = new ConstantLine();
            trendlineBorder.AxisValue = draggingTrendline.Point1.Argument;
            trendlineBorder.LineStyle.DashStyle = DashStyle.Dash;
            trendlineBorder.LineStyle.Thickness = 1;
            trendlineBorder.ShowInLegend = false;
            View.AxisX.ConstantLines.Add(trendlineBorder);
            chart.Capture = true;
            Cursor.Current = Cursors.SizeWE;
            UpdateControls();            
        }
        void chart_MouseUp(object sender, MouseEventArgs e) {
            if (View == null)
                return;
            if (draggingTrendline != null) {
                draggingTrendline.ExtrapolateToInfinity = extrapolateForNext;
                if (draggingTrendline.Point1.Argument.Equals(draggingTrendline.Point2.Argument))
                    View.Indicators.Remove(draggingTrendline);                
                else
                    chart.SetObjectSelection(draggingTrendline);
                View.AxisX.ConstantLines.Remove(trendlineBorder);
                draggingTrendline = null;                
            }
            chart.Capture = false;            
            Cursor.Current = Cursors.Default;
            UpdateControls();
        }
        void chart_MouseMove(object sender, MouseEventArgs e) {
            if (View == null)
                return;
            ChartHitInfo hitInfo = chart.CalcHitInfo(e.Location);
            if (draggingTrendline != null && e.Button == MouseButtons.Left) {
                if (hitInfo.InDiagram) {
                    DiagramCoordinates coords = Diagram.PointToDiagram(e.Location);
                    if (!coords.IsEmpty) {
                        draggingTrendline.Point2.Argument = CalculateArgument(coords.DateTimeArgument);
                        if ((DateTime)draggingTrendline.Point1.Argument > (DateTime)draggingTrendline.Point2.Argument) {
                            draggingTrendline.Point1.ValueLevel = EndValueLevel;
                            draggingTrendline.Point2.ValueLevel = StartValueLevel;
                        }
                        else {
                            draggingTrendline.Point1.ValueLevel = StartValueLevel;
                            draggingTrendline.Point2.ValueLevel = EndValueLevel;
                        }
                        trendlineBorder.AxisValue = draggingTrendline.Point2.Argument;
                    }
                }
            }
            else if (chart.SelectionMode == ElementSelectionMode.None)
                chart.SelectionMode = ElementSelectionMode.Single;
        }
        void chart_KeyDown(object sender, KeyEventArgs e) {
            if (e.KeyCode == Keys.Delete)
                DeleteSelectedTrendline();
        }        
        void btnDelete_Click(object sender, EventArgs e) {
           DeleteSelectedTrendline();           
        }
        void btnDeleteAll_Click(object sender, EventArgs e) {
            if (View != null)
                View.Indicators.Clear();
            selectedTrendline = null;
            UpdateControls();
        }
        void cbeStarValueLevel_SelectedIndexChanged(object sender, EventArgs e) {
            if (View != null)
                if (selectedTrendline == null)
                    startValueLevelForNext = StartValueLevel;
                else
                    selectedTrendline.Point1.ValueLevel = StartValueLevel;
        }
        void cbeEndValueLevel_SelectedIndexChanged(object sender, EventArgs e) {
            if (View != null)
                if (selectedTrendline == null)
                    endValueLevelForNext = EndValueLevel;
                else
                    selectedTrendline.Point2.ValueLevel = EndValueLevel;
        }
        void cheExtrapolateToInfinity_CheckedChanged(object sender, EventArgs e) {
            if (selectedTrendline == null)
                extrapolateForNext = cheExtrapolateToInfinity.Checked;
            else
                selectedTrendline.ExtrapolateToInfinity = cheExtrapolateToInfinity.Checked;
        }
        void ceColor_CustomDisplayText(object sender, CustomDisplayTextEventArgs e) {
            if ((Color)e.Value == Color.Empty)
                e.DisplayText = "Default";
        }
        void ceColor_EditValueChanged(object sender, EventArgs e) {
            if (selectedTrendline == null)
                colorForNext = ceColor.Color;
            else
                selectedTrendline.Color = ceColor.Color;
        }
        void ceColor_Properties_ButtonClick(object sender, ButtonPressedEventArgs e) {
            if (e.Button.Kind == ButtonPredefines.Delete)
                ceColor.Color = Color.Empty;
        }
        void chart_BoundDataChanged(object sender, EventArgs e) {
            ChartControl.Animate();
        }

        protected override void InitControls() {
            base.InitControls();
            FillDefaultTrendlines();
            FillComboBoxEditByValueLevels(cbeStarValueLevel, startValueLevelForNext);
            FillComboBoxEditByValueLevels(cbeEndValueLevel, endValueLevelForNext);
            ShowLabels = false;
            CurrentSeries.DataSource = MarketPrices.GetAppleStockPrices();
        }
        protected override void OnChartControlObjectHotTracked(object sender, HotTrackEventArgs e) {
            if (!(e.Object is TrendLine)) {
                if (!chart.Capture)
                    Cursor = Cursors.Default;
                e.Cancel = true;
            }
            else if (draggingTrendline == null)
                Cursor = Cursors.Hand;
            else
                e.Cancel = true;
        }
        protected override void OnChartControlObjectSelected(object sender, HotTrackEventArgs e) {
            if (e.HitInfo.InIndicator)
                selectedTrendline = e.HitInfo.Indicator as TrendLine;
            else
                selectedTrendline = null;
            if (selectedTrendline == null)
                e.Cancel = true;
            UpdateControls();
        }

        public override void UpdateControls() {
            base.UpdateControls();
            btnDelete.Enabled = selectedTrendline != null;
            if (View != null)
                btnDeleteAll.Enabled = View.Indicators.Count != 0;
            if (selectedTrendline == null) {
                cheExtrapolateToInfinity.Checked = extrapolateForNext;
                cbeStarValueLevel.SelectedItem = startValueLevelForNext;
                cbeEndValueLevel.SelectedItem = endValueLevelForNext;
                ceColor.Color = colorForNext;
            }
            else {
                cheExtrapolateToInfinity.Checked = selectedTrendline.ExtrapolateToInfinity;
                cbeStarValueLevel.SelectedItem = selectedTrendline.Point1.ValueLevel;
                cbeEndValueLevel.SelectedItem = selectedTrendline.Point2.ValueLevel;
                ceColor.Color = selectedTrendline.Color;
            }
        }
    }
}
